tensorflow_probability/python/experimental/mcmc/covariance_reducer.py

```diff
@@ -0,0 +1,185 @@
+        Default value: `None` (i.e., 'covariance_reducer').
+    """
+    self._parameters = dict(
+        shape=shape,
+        event_ndims=event_ndims,
+        dtype=dtype,
+        name=name or 'covariance_reducer'
+    )
+    self.strm = sample_stats.RunningCovariance(shape, event_ndims, dtype)
+
+  def initialize(self, initial_chain_state=None, initial_kernel_results=None):
+    """Initializes a `RunningCovarianceState` using previously defined metadata.
+
+    Args:
+      initial_chain_state: `Tensor` or Python `list` of `Tensor`s representing
+        the current state(s) of the Markov chain(s). For streaming covariance,
+        this argument has no influence on computation. Hence, by default, it
+        is `None`.
+      initial_kernel_results: A (possibly nested) `tuple`, `namedtuple` or
+        `list` of `Tensor`s representing internal calculations made in a related
+        `TransitionKernel`. For streaming covariance, this argument also has no
+        influence on computation; hence, it is also `None` by default.
+
+    Returns:
+      state: `RunningCovarianceState` representing a stream of no inputs.
+    """
+    # `initial_chain_state` not included as a sample for consistency
+    # with `sample_chain`
+    with tf.name_scope(
+        mcmc_util.make_name(self.name, 'covariance_reducer', 'initialize')):
+      return self.strm.initialize()
+
+  def one_step(
+      self,
+      sample,
+      current_reducer_state,
+      previous_kernel_results=None,
+      axis=None):
+    """Update the `current_reducer_state` with a new sample.
+
+    Args:
+      sample: Incoming sample with shape and dtype compatible with those
+        used to initialize the `CovarianceReducer` and the
+        `current_reducer_state`.
+      current_reducer_state: `RunningCovarianceState` representing the current
+        state of the running covariance.
+      previous_kernel_results: A (possibly nested) `tuple`, `namedtuple` or
+        `list` of `Tensor`s representing internal calculations made in a related
+        `TransitionKernel`. For streaming covariance, this argument has no
+        influence on computation; hence, it is `None` by default.
+      axis: If chunking is desired, this is an integer that specifies the axis
+        with chunked samples. For individual samples, set this to `None`. By
+        default, samples are not chunked (`axis` is None).
+
+    Returns:
+      new_state: `RunningCovarianceState` with updated running statistics.
+    """
+    with tf.name_scope(
+        mcmc_util.make_name(self.name, 'covariance_reducer', 'one_step')):
+      return self.strm.update(current_reducer_state, sample, axis=axis)
+
+  def finalize(self, final_state, ddof=0):
+    """Finalizes covariance calculation from the `final_state`.
+
+    final_state: `RunningCovarianceState` that represents the final state of
+        running statistics.
+    ddof: Requested dynamic degrees of freedom. For example, use `ddof=0`
```

Comment:
An alternative is to have the user pass in `finalize` kwargs in `sample_fold`. For example having

```
def sample_fold(..., **kwargs):
  ...
  reduction_result = nest.map_structure_up_to(
      reducers,
      lambda r, s: r.finalize(s, **kwargs),
      reducers,
      final_kernel_results.streaming_calculations,
      check_types=False)
  ...
``` 

That `sample_fold` use-case allows users to create multi-purposed reducers. For example, one covariance reducer is enough to support all ddof finalizes, instead of needing to make multiple. Should we opt for this alternative instead?

Reply:
As written, this seems to call for every reducer being able to accept any kwarg, so that a hypothetical ESSReducer's finalize would also have to accept (and ignore) ddof, to cover the situation where a user wants both covariance and ESS, and wants to pass a ddof argument to the covariance.

I'm also not sure what benefit you are looking for.  Reusing the same CovarianceReducer object across multiple calls to sample_fold, with different ddof arguments?  Why?  CovarianceReducer is stateless; making new ones is cheap computationally, and I'm not sure the potential usability savings is worth the mental complexity of plumbing arguments through sample_fold to finalize.  Open to being persuaded otherwise, however.

Reply:
Good points. I wanted to bring this possibility up since it popped into my head, but I think I'm now convinced that what I just described isn't really a good alternative.